### PR TITLE
Resnet encoders

### PR DESCRIPTION
**Pull request by ibadami**
I have managed to make a clean PR with this one. I think in the last ones I merged the upstream instead of rebasing it . 
To keep things cleaner, I also created a branch called resnet-encoders. Which now contains the previous bug fix in resnet 18 and additionally multiple resnet support.
I think it is better to use pre-trained resnet but we should also keep the resnet architecture in case one wants to train it from scratch for some reason best known to him.
I will shortly add the pretrained model support as well.

### CODE REVIEW

**Comment by meetshah1995** [comment]
Looks Good.

File: lr_scheduling.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        :param iter is a current iteration
+        :param lr_decay_iter how frequently decay occurs, default is 1
+        :param max_iter is number of maximum iterations
+        :param power is a polymomial power
+
+    """
+    if iter % lr_decay_iter or iter > max_iter:
+        return optimizer
+
+    for param_group in optimizer.param_groups:
```
Comment (saeedranjbar12) [inline lr_scheduling.py]: Lines of code after return does not make sense. I think return should be inserted after updating the learning rate.
